### PR TITLE
Upgrade to scala 2.12

### PR DESCRIPTION
**Pull request by adleong**

### CODE REVIEW

File: project/Base.scala [hunk 1]
```diff
@@ -47,7 +47,8 @@ class Base extends Build {
     organization := "io.buoyant",
     version := Git.version,
     homepage := Some(url("https://linkerd.io")),
-    scalaVersion in GlobalScope := "2.11.7",
+    scalaVersion in GlobalScope := "2.12.1",
```
Comment (zackangelo) [inline project/Base.scala]: Just a heads up, 2.12.2 is out now.

Comment (adleong) [inline project/Base.scala]: Thanks!  Even though it might not be strictly necessary, I think we're going to mirror the scala version that finagle uses for now.

**Comment by olix0r** [comment]
I believe finagle has 2.12 builds now, but we should check that all of our
dependencies are covered

File: finagle/h2/src/test/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransportTest.scala [hunk 2]
```diff
@@ -78,6 +78,7 @@ class Netty4StreamTransportTest extends FunSuite {
 
       eventually { assert(transport.onReset.isDefined) }
       assert(await(transport.onReset.liftToTry) == Throw(StreamError.Remote(rst)))
+      ()
     }
```
Comment (olix0r) [inline finagle/h2/src/test/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransportTest.scala]: So I understand... what does assert return now?

File: namerd/iface/interpreter-thrift/src/test/scala/io/buoyant/namerd/iface/ThriftNamerInterfaceTest.scala [hunk 3]
```diff
@@ -16,7 +16,7 @@ import org.scalatest.FunSuite
 class ThriftNamerInterfaceTest extends FunSuite with Awaits {
   import ThriftNamerInterface._
 
-  def retryIn() = 1.second
+  val retryIn = () => 1.second
```
Comment (olix0r) [inline namerd/iface/interpreter-thrift/src/test/scala/io/buoyant/namerd/iface/ThriftNamerInterfaceTest.scala]: so I understand, why is this necessary?

Comment (adleong) [inline namerd/iface/interpreter-thrift/src/test/scala/io/buoyant/namerd/iface/ThriftNamerInterfaceTest.scala]: `Eta-expansion of zero-argument method values is deprecated.`

It works if we replace usages of `retryIn` with `retryIn _` but I figured just val'ing up the function to begin with would be easier to understand.

File: finagle/h2/src/test/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransportTest.scala (hunk 2 above)
Comment (adleong) [inline finagle/h2/src/test/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransportTest.scala]: It returns [Assertion](http://doc.scalatest.org/3.0.1/index.html#org.scalatest.compatible.Assertion) which is just a marker trait.
